bundles/org.eclipse.equinox.p2.core/src/org/eclipse/equinox/internal/p2/core/ProvisioningAgent.java

```diff
@@ -50,42 +51,32 @@ public ProvisioningAgent() {
 	public Object getService(String serviceName) {
 		//synchronize so concurrent gets always obtain the same service
 		synchronized (agentServices) {
-			checkRunning();
 			Object service = agentServices.get(serviceName);
 			if (service != null) {
 				return service;
 			}
-			//attempt to get factory service from service registry
-			Collection<ServiceReference<IAgentServiceFactory>> refs;
-			try {
-				refs = context.getServiceReferences(IAgentServiceFactory.class,
-						String.format("(|(%s=%s)(p2.agent.servicename=%s))", //$NON-NLS-1$ use old property as fallback
-								IAgentServiceFactory.PROP_AGENT_SERVICE_NAME, serviceName, serviceName));
-			} catch (InvalidSyntaxException e) {
-				e.printStackTrace();
-				return null;
-			}
-			if (refs == null || refs.isEmpty()) {
-				return null;
-			}
-			ServiceReference<IAgentServiceFactory> firstRef = Collections.max(refs);
-			//track the factory so that we can automatically remove the service when the factory goes away
-			ServiceTracker<IAgentServiceFactory, Object> tracker = new ServiceTracker<>(context, firstRef, this);
-			tracker.open();
-			IAgentServiceFactory factory = (IAgentServiceFactory) tracker.getService();
-			if (factory == null) {
-				tracker.close();
-				return null;
-			}
-			service = factory.createService(this);
-			if (service == null) {
-				tracker.close();
-				return null;
-			}
-			registerService(serviceName, service);
-			trackers.put(firstRef, tracker);
-			return service;
 		}
+		ServiceTracker<IAgentServiceFactory, Object> tracker = trackers.computeIfAbsent(serviceName,
+				ref -> {
+					try {
+						if (stopped) {
+							return null;
+						}
+						Filter filter = context.createFilter(
+								String.format("(&(%s=%s)(|(%s=%s)(p2.agent.servicename=%s)))", //$NON-NLS-1$
+										Constants.OBJECTCLASS, IAgentServiceFactory.class.getName(), //
+										IAgentServiceFactory.PROP_AGENT_SERVICE_NAME, serviceName, //
+										serviceName)); // use old property as fallback
+						return new ServiceTracker<>(context, filter, trackerCustomizer);
+					} catch (InvalidSyntaxException e) {
+						throw new AssertionError(e);
```

Comment:
As `serviceName` is not validated syntax error may happen without any other error in agent code. `AssertionError` should only be thrown on internal errors. Either validate `serviceName` or throw `IllegalArgumentException`.